src/dex/teddyswap.ts

```diff
@@ -0,0 +1,233 @@
+import { LiquidityPool } from './models/liquidity-pool';
+import { BaseDataProvider } from '@providers/data/base-data-provider';
+import { Asset, Token } from './models/asset';
+import { BaseDex } from './base-dex';
+import { DatumParameters, DefinitionConstr, DefinitionField, PayToAddress, RequestConfig, SwapFee, UTxO } from '@app/types';
+import { DefinitionBuilder } from '@app/definition-builder';
+import { AddressType, DatumParameterKey } from '@app/constants';
+import { BaseApi } from '@dex/api/base-api';
+import pool from './definitions/teddyswap/pool';
+import order from './definitions/teddyswap/order';
+import { correspondingReserves, tokensMatch } from '..';
+
+export class TeddySwap extends BaseDex {
+
+    public static readonly identifier: string = 'TeddySwap';
+    public readonly api: BaseApi;
+
+    /**
+     * On-Chain constants.
+     */
+    public readonly orderAddress: string = 'addr1z99tz7hungv6furtdl3zn72sree86wtghlcr4jc637r2eadkp2avt5gp297dnxhxcmy6kkptepsr5pa409qa7gf8stzs0706a3';
+    public readonly poolAddress: string = 'addr1zy5th50h46anh3v7zdvh7ve6amac7k4h3mdfvt0p6czm8z9kp2avt5gp297dnxhxcmy6kkptepsr5pa409qa7gf8stzsxg8sx3';
+
+
+    constructor(requestConfig: RequestConfig = {}) {
+        super();
+    }
+
+    public async liquidityPoolAddresses(provider: BaseDataProvider): Promise<string[]> {
+        return Promise.resolve([this.poolAddress]);
+    }
+
+    async liquidityPools(provider: BaseDataProvider): Promise<LiquidityPool[]> {
+        const poolAddresses: string[] = await this.liquidityPoolAddresses(provider);
+
+        const addressPromises: Promise<LiquidityPool[]>[] = poolAddresses.map(async (address: string) => {
+            const utxos: UTxO[] = await provider.utxos(address);
+
+            return await Promise.all(
+                utxos.map(async (utxo: UTxO) => {
+                    return await this.liquidityPoolFromUtxo(provider, utxo);
+                })
+            ).then((liquidityPools: (LiquidityPool | undefined)[]) => {
+                return liquidityPools.filter((liquidityPool?: LiquidityPool) => {
+                    return liquidityPool !== undefined;
+                }) as LiquidityPool[];
+            });
+
+        });
+
+        return Promise.all(addressPromises).then((liquidityPools: Awaited<LiquidityPool[]>[]) => liquidityPools.flat());
+    }
+
+    public async liquidityPoolFromUtxo(provider: BaseDataProvider, utxo: UTxO): Promise<LiquidityPool | undefined> {
+        if (!utxo.datumHash) {
+            return Promise.resolve(undefined);
+        }
+
+        const relevantAssets = utxo.assetBalances.filter((assetBalance) => {
+            const assetName = (assetBalance.asset as Asset).assetName;
```

Comment:
```suggestion
            const assetName = assetBalance.asset === 'lovelace' ? 'lovelace' : assetBalance.assetName;
```

Reply:
: assetBalance.asset.assetName;

Is asset not required here?

Reply:
Yes it is! My comment was wrong but the code reflects the correct path now